src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala

```diff
@@ -81,11 +84,11 @@ object Boot extends IOApp with LazyLogging {
 
         _ <- dependencies.cloudExtensionsInitializer.onBoot(dependencies.samApplication)
 
-        binding <- IO.fromFuture(IO(Http().newServerAt("0.0.0.0", 8080).bind(dependencies.samRoutes.route))).onError {
+        httpServer = IO.fromFuture(IO(Http().newServerAt("0.0.0.0", 8080).bind(dependencies.samRoutes.route))).onError {
           case t: Throwable => IO(logger.error("FATAL - failure starting http server", t)) *> IO.raiseError(t)
-        }
-        _ <- IO.fromFuture(IO(binding.whenTerminated))
-        _ <- IO(system.terminate())
+        }.void
+        allStreams = fs2.Stream.eval(httpServer) :: dependencies.cryptominingSubscriberProcess
+        _ <- fs2.Stream.emits(allStreams).covary[IO].parJoin(allStreams.length).compile.drain
```

Comment:
not knowing a lot about streams, could you summarize what is happening here? (and maybe I should read up on Streams)

Reply:
This is a way to run various processes concurrently. In this case, it's the cryptomining subscriber, and the main sam http server.

For your question above, because the Streams here are expected to be all `Stream[IO, Unit]`, hence we're using `void` up there. Without `void`, it'll return `IO[Http.ServerBinding]`, which can't be combined with other Streams for running concurrently.

Reply:
OH ok. I think I also missed the `eval` in `fs2.Stream.eval(httpServer)` -- which would allow the stream to evaluate the httpServer IO.